clippy_lints/src/redundant_clone.rs

```diff
@@ -135,43 +119,23 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for RedundantClone {
                 continue;
             }
 
-            // _1 in MIR `{ _2 = &_1; clone(move _2); }` or `{ _2 = _1; to_path_buf(_2); }
-            let cloned = if let Some(referent) = bbdata
-                .statements
-                .iter()
-                .rev()
-                .filter_map(|stmt| {
-                    if let mir::StatementKind::Assign(mir::Place::Local(local), v) = &stmt.kind {
-                        if *local == arg {
-                            if from_deref {
-                                // `r` is already a reference.
-                                if let mir::Rvalue::Use(mir::Operand::Copy(mir::Place::Local(r))) = **v {
-                                    return Some(r);
-                                }
-                            } else if let mir::Rvalue::Ref(_, _, mir::Place::Local(r)) = **v {
-                                return Some(r);
-                            }
-                        }
-                    }
-
-                    None
-                })
-                .next()
-            {
-                referent
-            } else {
-                continue;
-            };
+            // _1 in MIR `{ _2 = &_1; clone(move _2); }` or `{ _2 = _1; to_path_buf(_2); } (from_deref)
+            // In case of `from_deref`, `arg` is already a reference since it is `deref`ed in the previous
+            // block.
+            let cloned = unwrap_or_continue!(find_stmt_assigns_to(arg, from_borrow, bbdata.statements.iter().rev()));
 
             // _1 in MIR `{ _2 = &_1; _3 = deref(move _2); } -> { _4 = _3; to_path_buf(move _4); }`
             let referent = if from_deref {
                 let ps = mir.predecessors_for(bb);
+                if ps.len() != 1 {
+                    continue;
+                }
+                let pred_terminator = unwrap_or_continue!(&mir[ps[0]].terminator);
```

Comment:
ah, this should never be `None` in MIR you get from the compiler. use the `terminator` method to get the value directly.